llvm/tools/sycl-post-link/ModuleSplitter.h

```diff
@@ -34,66 +35,122 @@ enum IRSplitMode {
+    SyclEsimdSplitStatus HasESIMD = SyclEsimdSplitStatus::SYCL_AND_ESIMD;
+    // Whether any of the EPs use double GRF mode.
+    bool UsesDoubleGRF = false;
+    // Scope represented by EPs in a group
+    EntryPointsGroupScope Scope = Scope_Global;
+
+    Properties merge(const Properties &Other) const {
+      Properties Res;
+      Res.HasESIMD = HasESIMD == Other.HasESIMD
+                         ? HasESIMD
+                         : SyclEsimdSplitStatus::SYCL_AND_ESIMD;
+      Res.UsesDoubleGRF = UsesDoubleGRF || Other.UsesDoubleGRF;
+      // Scope remains global
+      return Res;
+    }
+  };
 
-public:
-  EntryPointVec Functions;
+  StringRef GroupId;
+  EntryPointSet Functions;
+  Properties Props;
 
   EntryPointGroup(StringRef GroupId = "") : GroupId(GroupId) {}
-  EntryPointGroup(StringRef GroupId, EntryPointVec &&Functions)
+  EntryPointGroup(StringRef GroupId, EntryPointSet &&Functions)
       : GroupId(GroupId), Functions(std::move(Functions)) {}
+  EntryPointGroup(StringRef GroupId, EntryPointSet &&Functions,
+                  const Properties &Props)
+      : GroupId(GroupId), Functions(std::move(Functions)), Props(Props) {}
+
+  // Tells if this group has only ESIMD entry points.
+  bool isEsimd() const {
+    return Props.HasESIMD == SyclEsimdSplitStatus::ESIMD_ONLY;
+  }
+  // Tells if this group has only SYCL entry points.
+  bool isSycl() const {
+    return Props.HasESIMD == SyclEsimdSplitStatus::SYCL_ONLY;
+  }
+  // Tells if some entry points use double GRF mode.
+  bool isDoubleGRF() const { return Props.UsesDoubleGRF; }
 
-  const StringRef getId() const { return GroupId; }
-  // Tells if this group has only ESIMD entry points (based on GroupId).
-  bool isEsimd() const;
-  // Tells if this group has only SYCL entry points (based on GroupId).
-  bool isSycl() const;
   void saveNames(std::vector<std::string> &Dest) const;
   void rebuildFromNames(const std::vector<std::string> &Names, const Module &M);
   void rebuild(const Module &M);
 };
 
 using EntryPointGroupVec = std::vector<EntryPointGroup>;
 
-enum class SyclEsimdSplitStatus { SYCL_ONLY, ESIMD_ONLY, SYCL_AND_ESIMD };
-
+// Annotates an llvm::Module with information necessary to perform and track
+// result of device code (llvm::Module instances) splitting:
+// - entry points of the module determined e.g. by a module splitter, as well
+//   as information about entry point origin (e.g. result of a scoped split)
+// - its properties, such as whether it has specialization constants uses
+// It also provides convenience functions for entry point set transformation
+// between llvm::Function object and string representations.
 class ModuleDesc {
   std::unique_ptr<Module> M;
   EntryPointGroup EntryPoints;
+  bool IsTopLevel = false;
 
 public:
   struct Properties {
-    SyclEsimdSplitStatus HasEsimd = SyclEsimdSplitStatus::SYCL_AND_ESIMD;
-    bool SpecConstsMet = true;
+    bool SpecConstsMet = false;
```

Comment:
Why do we need a struct here? Do we expect this to expand beyond a single field?

Reply:
Can be addressed later.